README
#=GCal4Ruby
#
#==Introduction
#GCal4Ruby is a full featured wrapper for the google calendar API.  GCal4Ruby implements
#all of the functionality available through the Google Calnedar API, including permissions,
#attendees, reminders and event recurrence.  
#
#==Author and Contact Information
#GCal4Ruby was created and is maintained by {Mike Reich}[mailto:mike@seabourneconsulting.com] 
#and is licenses under the GPL v2.  Feel free to use and update, but be sure to contribute your
#code back to the project and attribute as required by the license.
#===Website
#http://rubyforge.org/projects/gcal4ruby/
#
#==Description
#GCal4Ruby has three major components: the service, calendar and event objects.  Each service
#has many calendars, which in turn have many events.  Each service is the representation of a
#google account, and thus must be successfully authenticated using valid Google Calendar
#account credentials.  
#
#==Examples
#Below are some common usage examples.  For more examples, check the documentation.
#===Service
#1. Authenticate
#    service = Service.new
#    service.authenticate("user@gmail.com", "password")
#
#2. Get Calendar List
#    calendars = service.calendars
#
#===Calendar
#All usages assume a successfully authenticated Service.
#1. Create a new Calendar
#    cal = Calendar.new(service)
#
#2. Find an existing Calendar
#    cal = Calendar.find(service, "New Calendar", :first)
#
#3. Find all calendars containing the search term
#    cal = Calendar.find(service, "Soccer Team")
#
#4. Find a calendar by ID
#    cal = Calendar.find(service, id, :first)
#===Event
#All usages assume a successfully authenticated Service and valid Calendar.
#1. Create a new Event
#    event = Event.new(calendar)
#    event.title = "Soccer Game"
#    event.start = Time.parse("12-06-2009 at 12:30 PM")
#    event.end = Time.parse("12-06-2009 at 1:30 PM")
#    event.where = "Merry Playfields"
#    event.save
#
#2. Find an existing Event
#    event = Event.find(cal, "Soccer Game", :first)
#
#3. Find all events containing the search term
#    event = Event.find(cal, "Soccer Game")
#
#4. Create a recurring event for every saturday
#    event = Event.new(calendar)
#    event.title = "Baseball Game"
#    event.where = "Municipal Stadium"
#    event.recurrence = Recurrence.new
#    event.recurrence.start = Time.parse("13-06-2009 at 4:30 PM")
#    event.recurrence.end = Time.parse("13-06-2009 at 6:30 PM")
#    event.recurrence.frequency = {"weekly" => ["SA"]}
#    event.save 
#
#5. Create an event with a 15 minute email reminder
#    event = Event.new(calendar)
#    event.title = "Dinner with Kate"
#    event.start = Time.parse("20-06-2009 at 5 pm")
#    event.end = Time.parse("20-06-209 at 8 pm")
#    event.where = "Luigi's"
#    event.reminder = {:minutes => 15, :method => 'email'}
#    event.save
#
#6. Create an event with attendees
#    event = Event.new(calendar)
#    event.title = "Dinner with Kate"
#    event.start = Time.parse("20-06-2009 at 5 pm")
#    event.end = Time.parse("20-06-209 at 8 pm")
#    event.attendees => {:name => "Kate", :email => "kate@gmail.com"}
#    event.save


CHANGELOG
#=CHANGELOG
#==version 0.2.6
#* Added fix for updated google calendar XML per http://cookingandcoding.wordpress.com/2009/06/08/new-ruby-google-calendar-api-gem-gcal4ruby/#comment-183
#==version 0.2.5
#* Added calendar color support to to_iframe methods in Calendar and Service.
#==version 0.2.4
#* Fixed bug with ACL check in Calendar#load
#==version 0.2.3
#* Implemented to_iframe method for calendars and services to output embeddable iframe text.
#* Added switch to turn off ACL check for public calendars.  Can increase effeciency if turned off.
#==version 0.2.2
#* Fixed URL encoding problem in Event.find method.
#* cleaned up Event.find method to allow for finding events by id
#* updated Calendar.find method to add params hash
#* added 'published', 'updated', and 'edited' attributes
#==version 0.2.1
#* fixed Event.find calendar specification
#==version 0.2.0
#* Fixed redirect URI query params problem
#* Updated syntax for finding events to include most google api parameters,  Non-backwards compatible.
#==version 0.1.4
#* Added additional search criteria for Event.find
#==version 0.1.3
#* Added support for authenticating with Google Hosted Apps accounts
#* Added flag to indicate whether a calendar is editable
#* Added handling to gracefully throw error when trying to create event on a non-editable (shared) calendar
#==version 0.1.2
#* Fixed to_xml dump problem with hidden and selected attributes
#==version 0.1.1
#* Added all_day indicator to event to indicate an all day event
#==version 0.1.0
#* Initial Version

test/unit.rb
#!/usr/bin/ruby

require 'rubygems'
require 'gcal4ruby'
include GCal4Ruby

@service = Service.new
@username = nil
@password = nil

def tester
  if ARGV.include?("-d")
      @service.debug = true
  end
  ARGV.each do |ar|
    if ar.match("username=")
      @username = ar.gsub("username=", "")
    end
    if ar.match("password=")
      @password = ar.gsub("password=", "")
    end
  end
  service_test
  calendar_test
  event_test
end

def service_test
  puts "---Starting Service Test---"
  puts "1. Authenticate"
  if @service.authenticate(@username, @password)
    successful
  else
    failed
  end
  
  puts "2. Calendar List"
  cals = @service.calendars
  if cals
    successful "Calendars for this Account:"
    cals.each do |cal|
      puts cal.title
    end
  else
    failed
  end
end

def calendar_test
  puts "---Starting Calendar Test---"
  
  puts "1. Create Calendar"
  cal = Calendar.new(@service)
  cal.title = "test calendar"+Time.now.to_s
  puts "Calender exists = "+cal.exists?.to_s
  if cal.save
    successful cal.to_xml
  else
    failed
  end
  
  puts "2. Edit Calendar"
  cal.title = "renamed title"
  if cal.save
    successful cal.to_xml
  else
    puts "Test 2 Failed"
  end
  
  puts "3. Find Calendar by ID"
  c = Calendar.find(@service, cal.id)
  if c.title == cal.title
    successful
  else
    failed "#{c.title} not equal to #{cal.title}"
  end
  
  puts "4. Delete Calendar"
  if cal.delete and not cal.title
    successful
  else
    failed
  end
end

def event_test
  puts "---Starting Event Test---"
  
  puts "1. Create Event"
  event = Event.new(@service.calendars[0])
  event.title = "Test Event"
  event.content = "Test event content"
  event.start = Time.now+1800
  event.end = Time.now+5400
  if event.save
    successful event.to_xml
  else
    failed
  end
  
  puts "2. Edit Event"
  event.title = "Edited title"
  if event.save
    successful event.to_xml
  else
    failed
  end
  
  puts "3. Reload Event"
  if event.reload
    successful
  end
  
  puts "4. Find Event by id"
  e = Event.find(@service.calendars[0], event.id)
  if e.title == event.title
    successful
  else
    failed "Found event doesn't match existing event"
  end
  
  puts "5. Delete Event"
  if event.delete
    successful 
  else
    failed
  end
end

def failed(m = nil)
  puts "Test Failed"
  puts m if m
  exit()
end

def successful(m = nil)
  puts "Test Successful"
  puts m if m
end

tester

lib/gcal4ruby/base.rb
require "rexml/document"
require "cgi"
require "uri"
require "net/http"
require "net/https"
require "open-uri"
require "nkf"
require "time"

Net::HTTP.version_1_2

# GCal4Ruby is a full featured wrapper for the google calendar API

# =Usage:

module GCal4Ruby
  
  CALENDAR_XML = "<entry xmlns='http://www.w3.org/2005/Atom' 
       xmlns:gd='http://schemas.google.com/g/2005' 
       xmlns:gCal='http://schemas.google.com/gCal/2005'>
  <title type='text'></title>
  <summary type='text'></summary>
  <gCal:timezone value=''></gCal:timezone>
  <gCal:hidden value=''></gCal:hidden>
  <gCal:color value=''></gCal:color>
  <gd:where rel='' label='' valueString=''></gd:where>
  </entry>"
  
  ACL_XML = "<entry xmlns='http://www.w3.org/2005/Atom' xmlns:gAcl='http://schemas.google.com/acl/2007'>
                      <category scheme='http://schemas.google.com/g/2005#kind' term='http://schemas.google.com/acl/2007#accessRule'/>
                       <gAcl:scope type='default'></gAcl:scope>
                       <gAcl:role value=''></gAcl:role>
                    </entry>"
                    
  EVENT_XML = "<entry xmlns='http://www.w3.org/2005/Atom' xmlns:gd='http://schemas.google.com/g/2005'>
  <category scheme='http://schemas.google.com/g/2005#kind' term='http://schemas.google.com/g/2005#event'></category>
  <title type='text'></title>
  <content type='text'></content>
  <gd:transparency value=''></gd:transparency>
  <gd:eventStatus value=''></gd:eventStatus>
  <gd:where valueString=''></gd:where>
  <gd:when startTime='' endTime=''></gd:when>
</entry>
  "
  
  class AuthenticationFailed < StandardError; end #:nodoc: all

  class NotAuthenticated < StandardError; end
    
  class InvalidService < StandardError; end
    
  class HTTPPostFailed < StandardError; end
    
  class HTTPPutFailed < StandardError; end
    
  class HTTPGetFailed < StandardError; end
    
  class HTTPDeleteFailed < StandardError; end
    
  class CalendarSaveFailed < StandardError; end
  
  class EventSaveFailed < StandardError; end
    
  class RecurrenceValueError < StandardError; end
    
  class CalendarNotEditable < StandardError; end
    
  class QueryParameterError < StandardError; end

  #The ProxyInfo class contains information for configuring a proxy connection

  class ProxyInfo
    attr_accessor :address, :port, :username, :password
    @address = nil
    @port = nil
    @username = nil
    @password = nil

    #The initialize function accepts four variables for configuring the ProxyInfo object.  
    #The proxy connection is initiated using the builtin Net::HTTP proxy support.

    def initialize(address, port, username=nil, password=nil)
      @address = address
      @port = port
      @username = username
      @password = password
    end
  end
  
  #The Base class includes the basic HTTP methods for sending and receiving 
  #messages from the Google Calendar API.  You shouldn't have to use this class 
  #directly, rather access the functionality through the Service subclass.

  class Base
    AUTH_URL = "https://www.google.com/accounts/ClientLogin"
    CALENDAR_LIST_FEED = "http://www.google.com/calendar/feeds/default/allcalendars/full"
    @proxy_info = nil
    @auth_token = nil
    @debug = false

    #Contains the ProxyInfo object for using a proxy server
    attr_accessor :proxy_info
    
    #If set to true, debug will dump all raw HTTP requests and responses
    attr_accessor :debug
    
    # Sends an HTTP POST request.  The header should be a hash of name/value pairs.  
    # Returns the Net::HTTPResponse object on succces, or raises the appropriate
    # error if a non 20x response code is received.
    def send_post(url, content, header=nil)
      header = auth_header(header)
      ret = nil
      location = URI.parse(url)
      https = get_http_object(location)
      puts "url = "+url if @debug
      if location.scheme == 'https'
        puts "SSL True" if @debug
      	https.use_ssl = true
      	https.verify_mode = OpenSSL::SSL::VERIFY_NONE
      end
      puts "Starting post\nHeader: #{header}\nContent: #{content}" if @debug
      https.start do |http|
        ret = http.post(location.to_s, content, header)
      end
      while ret.is_a?(Net::HTTPRedirection)
        puts "Redirect recieved, resending post" if @debug
        https.start do |http|
          ret = http.post(ret['location'], content, header)
        end
      end
      if ret.is_a?(Net::HTTPSuccess)
        puts "20x response recieved\nResponse: \n"+ret.read_body if @debug
        return ret
      else
        puts "invalid response received: "+ret.code if @debug
        raise HTTPPostFailed, ret.body
      end
    end
    
    # Sends an HTTP PUT request.  The header should be a hash of name/value pairs.  
    # Returns the Net::HTTPResponse object on succces, or raises the appropriate
    # error if a non 20x response code is received.
    def send_put(url, content, header=nil)
      header = auth_header(header)
      ret = nil
      location = URI.parse(url)
      https = get_http_object(location)
      puts "url = "+url if @debug
      if location.scheme == 'https'
        puts "SSL True" if @debug
        https.use_ssl = true
        https.verify_mode = OpenSSL::SSL::VERIFY_NONE
      end
      puts "Starting post\nHeader: #{header}\nContent: #{content}" if @debug
      https.start do |http|
        ret = http.put(location.to_s, content, header)
      end
      while ret.is_a?(Net::HTTPRedirection)
        puts "Redirect recieved, resending post" if @debug
        https.start do |http|
          ret = http.put(ret['location'], content, header)
        end
      end
      if ret.is_a?(Net::HTTPSuccess)
        puts "20x response recieved\nResponse: \n"+ret.read_body if @debug
        return ret
      else
        puts "invalid response received: "+ret.code if @debug
        raise HTTPPutFailed, ret.body
      end
    end

    # Sends an HTTP GET request.  The header should be a hash of name/value pairs.  
    # Returns the Net::HTTPResponse object on succces, or raises the appropriate
    # error if a non 20x response code is received.
    def send_get(url, header = nil)
      header = auth_header(header)
      ret = nil
      location = URI.parse(url)
      http = get_http_object(location)
      puts "url = "+url if @debug
      if location.scheme == 'https'
        puts "SSL True" if @debug
        https.use_ssl = true
        https.verify_mode = OpenSSL::SSL::VERIFY_NONE
      end
      puts "Starting post\nHeader: #{header}\n" if @debug
      http.start do |http|
        ret = http.get(location.to_s, header)
      end
      while ret.is_a?(Net::HTTPRedirection)
        puts "Redirect recieved, resending get to #{ret['location']}" if @debug
    	  http.start do |http|
    	    ret = http.get(ret['location'], header)
    	  end
      end
      if ret.is_a?(Net::HTTPSuccess)
        puts "20x response recieved\nResponse: \n"+ret.read_body if @debug
        return ret
      else
        puts "Error recieved, resending get" if @debug
        raise HTTPGetFailed, ret.body
      end
    end

    # Sends an HTTP DELETE request.  The header should be a hash of name/value pairs.  
    # Returns the Net::HTTPResponse object on succces, or raises the appropriate
    # error if a non 20x response code is received.
    def send_delete(url, header = nil)
      header = auth_header(header)
      ret = nil
      location = URI.parse(url)
      https = get_http_object(location)
      puts "url = "+url if @debug
      if location.scheme == 'https'
        puts "SSL True" if @debug
        https.use_ssl = true
        https.verify_mode = OpenSSL::SSL::VERIFY_NONE
      end
      puts "Starting post\nHeader: #{header}\n" if @debug
      https.start do |http|
        ret = http.delete(location.to_s, header)
      end
      while ret.is_a?(Net::HTTPRedirection)
        puts "Redirect recieved, resending post" if @debug
        https.start do |http|
          ret = http.delete(ret['location'], header)
        end
      end
      if ret.is_a?(Net::HTTPSuccess)
        puts "20x response recieved\nResponse: \n"+ret.read_body if @debug
        return true
      else
        puts "invalid response received: "+ret.code if @debug
        raise HTTPDeleteFailed, ret.body
      end
    end

    private

    def get_http_object(location)
      if @proxy_info and @proxy_info.address
	     return Net::HTTP.new(location.host, location.port, @proxy_info.address, @proxy_info.port, @proxy_info.username, @proxy_info.password)
      else
	     return Net::HTTP.new(location.host, location.port)
      end      
    end

    def auth_header(header)
      if @auth_token
      	if header
      	  header.merge!({'Authorization' => "GoogleLogin auth=#{@auth_token}", "GData-Version" => "2.1"})
      	else 
      	  header = {'Authorization' => "GoogleLogin auth=#{@auth_token}", "GData-Version" => "2.1"}
      	end
      end
      return header
    end
  end
end

lib/gcal4ruby/event.rb
require 'gcal4ruby/recurrence'

module GCal4Ruby
  #The Event Class represents a remote event in calendar. 
  #
  #=Usage
  #All usages assume a successfully authenticated Service and valid Calendar.
  #1. Create a new Event
  #    event = Event.new(calendar)
  #    event.title = "Soccer Game"
  #    event.start = Time.parse("12-06-2009 at 12:30 PM")
  #    event.end = Time.parse("12-06-2009 at 1:30 PM")
  #    event.where = "Merry Playfields"
  #    event.save
  #
  #2. Find an existing Event
  #    event = Event.find(cal, "Soccer Game", {:scope => :first})
  #
  #3. Find all events containing the search term
  #    event = Event.find(cal, "Soccer Game")
  #
  #4. Create a recurring event for every saturday
  #    event = Event.new(calendar)
  #    event.title = "Baseball Game"
  #    event.where = "Municipal Stadium"
  #    event.recurrence = Recurrence.new
  #    event.recurrence.start = Time.parse("13-06-2009 at 4:30 PM")
  #    event.recurrence.end = Time.parse("13-06-2009 at 6:30 PM")
  #    event.recurrence.frequency = {"weekly" => ["SA"]}
  #    event.save 
  #
  #5. Create an event with a 15 minute email reminder
  #    event = Event.new(calendar)
  #    event.title = "Dinner with Kate"
  #    event.start = Time.parse("20-06-2009 at 5 pm")
  #    event.end = Time.parse("20-06-209 at 8 pm")
  #    event.where = "Luigi's"
  #    event.reminder = {:minutes => 15, :method => 'email'}
  #    event.save
  #
  #6. Create an event with attendees
  #    event = Event.new(calendar)
  #    event.title = "Dinner with Kate"
  #    event.start = Time.parse("20-06-2009 at 5 pm")
  #    event.end = Time.parse("20-06-209 at 8 pm")
  #    event.attendees => {:name => "Kate", :email => "kate@gmail.com"}
  #    event.save
  #
  #After an event object has been created or loaded, you can change any of the 
  #attributes like you would any other object.  Be sure to save the event to write changes
  #to the Google Calendar service.
  class Event
    #The event title
    attr_accessor :title
    #The content for the event
    attr_accessor :content
    #The location of the event
    attr_accessor :where
    #A flag for whether the event show as :free or :busy
    attr_accessor :transparency
    #A flag indicating the status of the event.  Values can be :confirmed, :tentative or :cancelled
    attr_accessor :status
    #The unique event ID
    attr_accessor :id
    #Flag indicating whether it is an all day event
    attr_accessor :all_day
    
    @attendees
    
    #The event start time
    attr_reader :start
    #The event end time
    attr_reader :end
    #The reminder settings for the event, returned as a hash
    attr_reader :reminder
    #The date the event was created
    attr_reader :published
    #The date the event was last updated
    attr_reader :updated
    #The date the event was last edited
    attr_reader :edited
    
    #Sets the reminder options for the event.  Parameter must be a hash containing one of 
    #:hours, :minutes and :days, which are simply the number of each before the event start date you'd like to 
    #receive the reminder.  
    #
    #:method can be one of the following:
    #- <b>'alert'</b>: causes an alert to appear when a user is viewing the calendar in a browser
    #- <b>'email'</b>: sends the user an email message
    def reminder=(r)
      @reminder = r
    end
    
    #Returns the current event's Recurrence information
    def recurrence
      @recurrence
    end
    
    #Returns an array of the current attendees
    def attendees
      @attendees
    end
    
    #Accepts an array of email address/name pairs for attendees.  
    #  [{:name => 'Mike Reich', :email => 'mike@seabourneconsulting.com'}]
    #The email address is requried, but the name is optional
    def attendees=(a)
      if a.is_a?(Array)
        @attendees = a
      else
        raise "Attendees must be an Array of email/name hash pairs"
      end
    end
    
    #Sets the event's recurrence information to a Recurrence object.  Returns the recurrence if successful,
    #false otherwise
    def recurrence=(r)
      if r.is_a?(Recurrence)
        r.event = self
        @recurrence = r
      else
        return false
      end
    end
    
    #Returns a duplicate of the current event as a new Event object
    def copy()
      e = Event.new()
      e.load(to_xml)
      e.calendar = @calendar
      return e
    end
    
    #Sets the start time of the Event.  Must be a Time object or a parsable string representation
    #of a time.
    def start=(str)
      if str.is_a?String
        @start = Time.parse(str)      
      elsif str.is_a?Time
        @start = str
      else
        raise "Start Time must be either Time or String"
      end
    end
    
    #Sets the end time of the Event.  Must be a Time object or a parsable string representation
    #of a time.
    def end=(str)
      if str.is_a?String
        @end = Time.parse(str)      
      elsif str.is_a?Time
        @end = str
      else
        raise "End Time must be either Time or String"
      end
    end
    
    #Deletes the event from the Google Calendar Service.  All values are cleared.
    def delete
        if @exists    
          if @calendar.service.send_delete(@edit_feed, {"If-Match" => @etag})
            @exists = false
            @deleted = true
            @title = nil
            @content = nil
            @id = nil
            @start = nil
            @end = nil
            @transparency = nil
            @status = nil
            @where = nil
            return true
          else
            return false
          end
        else
          return false
        end
    end
    
    #Creates a new Event.  Accepts a valid Calendar object.
    def initialize(calendar)
      if not calendar.editable
        raise CalendarNotEditable
      end
      super()
      @xml = EVENT_XML
      @calendar = calendar
      @title = nil
      @content = nil
      @start = nil
      @end = nil
      @where = nil
      @transparency = "http://schemas.google.com/g/2005#event.opaque"
      @status = "http://schemas.google.com/g/2005#event.confirmed"
      @attendees = []
      @reminder = nil
      @all_day = false
    end
    
    #If the event does not exist on the Google Calendar service, save creates it.  Otherwise
    #updates the existing event data.  Returns true on success, false otherwise.
    def save
      if @deleted
        return false
      end
      if @exists 
        ret = @calendar.service.send_put(@edit_feed, to_xml, {'Content-Type' => 'application/atom+xml', "If-Match" => @etag})
      else
        ret = @calendar.service.send_post(@calendar.event_feed, to_xml, {'Content-Type' => 'application/atom+xml'})
      end
      if !@exists
        if load(ret.read_body)
          return true
        else
          raise EventSaveFailed
        end
      end
      reload
      return true
    end
    
    #Returns an XML representation of the event.
    def to_xml()
      xml = REXML::Document.new(@xml)
      xml.root.elements.each(){}.map do |ele|
        case ele.name
        when 'id'
          ele.text = @id
        when "title"
          ele.text = @title
        when "content"
          ele.text = @content
        when "when"
          if not @recurrence
            ele.attributes["startTime"] = @all_day ? @start.strftime("%Y-%m-%d") : @start.xmlschema
            ele.attributes["endTime"] = @all_day ? @end.strftime("%Y-%m-%d") : @end.xmlschema
            set_reminder(ele)
          else
            if not @reminder
              xml.root.delete_element("/entry/gd:when")
              xml.root.add_element("gd:recurrence").text = @recurrence.to_s
            else
              ele.delete_attribute('startTime')
              ele.delete_attribute('endTime')
              set_reminder(ele)  
            end
          end
        when "eventStatus"
          ele.attributes["value"] = case @status
            when :confirmed
              "http://schemas.google.com/g/2005#event.confirmed"
            when :tentative
              "http://schemas.google.com/g/2005#event.tentative"
            when :cancelled
              "http://schemas.google.com/g/2005#event.canceled"
            else
              "http://schemas.google.com/g/2005#event.confirmed"
          end
        when "transparency"
          ele.attributes["value"] = case @transparency
              when :free
                "http://schemas.google.com/g/2005#event.transparent"
              when :busy
                "http://schemas.google.com/g/2005#event.opaque"
              else
                "http://schemas.google.com/g/2005#event.opaque"
            end
        when "where"
          ele.attributes["valueString"] = @where
        end
      end        
      if not @attendees.empty?
        @attendees.each do |a|
          xml.root.add_element("gd:who", {"email" => a[:email], "valueString" => a[:name], "rel" => "http://schemas.google.com/g/2005#event.attendee"})
        end
      end
      xml.to_s
    end
    
    #Loads the event info from an XML string.
    def load(string)
      @xml = string
      @exists = true
      xml = REXML::Document.new(string)
      @etag = xml.root.attributes['etag']
      xml.root.elements.each(){}.map do |ele|
          case ele.name
             when 'updated'
                @updated = ele.text
             when 'published'
                @published = ele.text
             when 'edited'
                @edited = ele.text
             when 'id'
                @id, @edit_feed = ele.text
             when 'title'
                @title = ele.text
              when 'content'
                @content = ele.text
              when "when"
                @start = Time.parse(ele.attributes['startTime'])
                @end = Time.parse(ele.attributes['endTime'])
                ele.elements.each("gd:reminder") do |r|
                  @reminder = {:minutes => r.attributes['minutes'] ? r.attributes['minutes'] : 0, :hours => r.attributes['hours'] ? r.attributes['hours'] : 0, :days => r.attributes['days'] ? r.attributes['days'] : 0, :method => r.attributes['method'] ? r.attributes['method'] : ''}
                end
              when "where"
                @where = ele.attributes['valueString']
              when "link"
                if ele.attributes['rel'] == 'edit'
                  @edit_feed = ele.attributes['href']
                end
              when "who"
                if ele.attributes['rel'] == "http://schemas.google.com/g/2005#event.attendee"
                n = {}
                ele.attributes.each do |name, value|
                    case name
                      when "email"
                        n[:email] = value
                      when "valueString"
                        n[:name] = value
                    end
                  end                
               @attendees << n
               end
              when "eventStatus"
              case ele.attributes["value"] 
                when "http://schemas.google.com/g/2005#event.confirmed"
                 @status =  :confirmed
                when "http://schemas.google.com/g/2005#event.tentative"
                  @status = :tentative
                when "http://schemas.google.com/g/2005#event.confirmed"
                  @status = :cancelled
              end
            when "transparency"
               case ele.attributes["value"]
                  when "http://schemas.google.com/g/2005#event.transparent" 
                    @transparency = :free
                  when "http://schemas.google.com/g/2005#event.opaque"
                    @transparency = :busy
                end
            end      
        end
    end
    
    #Reloads the event data from the Google Calendar Service.  Returns true if successful,
    #false otherwise.
    def reload
      t = Event.find(@calendar, @id)
      if t
        if load(t.to_xml)
         return true
        else
         return false
        end
      else
        return false
      end
    end
    
    #Finds the event that matches a query term in the event title or description.
    #  
    #'query' is a string to perform the search on or an event id.
    # 
    #The params hash can contain the following hash values
    #* *scope*: may be :all or :first, indicating whether to return the first record found or an array of all records that match the query.  Default is :all.
    #* *range*: a hash including a :start and :end time to constrain the search by
    #* *max_results*: an integer indicating the number of results to return.  Default is 25.
    #* *sort_order*: either 'ascending' or 'descending'.
    #* *single_events*: either 'true' to return all recurring events as a single entry, or 'false' to return all recurring events as a unique event for each recurrence.
    #* *ctz*: the timezone to return the event times in
    def self.find(calendar, query = '', params = {})
      query_string = ''
      
      begin 
        test = URI.parse(query).scheme
      rescue Exception => e
        test = nil
      end
      
      if test
        puts "id passed, finding event by id" if calendar.service.debug
        es = calendar.service.send_get("http://www.google.com/calendar/feeds/#{calendar.id}/private/full")
        REXML::Document.new(es.read_body).root.elements.each("entry"){}.map do |entry|
          puts "element  = "+entry.name if calendar.service.debug
          id = ''
          entry.elements.each("id") {|v| id = v.text}
          puts "id = #{id}" if calendar.service.debug
          if id == query
            entry.attributes["xmlns:gCal"] = "http://schemas.google.com/gCal/2005"
            entry.attributes["xmlns:gd"] = "http://schemas.google.com/g/2005"
            entry.attributes["xmlns:app"] = "http://www.w3.org/2007/app" #Per http://twitter.com/mgornick
            entry.attributes["xmlns"] = "http://www.w3.org/2005/Atom"
            event = Event.new(calendar)
            event.load("<?xml version='1.0' encoding='UTF-8'?>#{entry.to_s}")
            return event
          end
        end
      end

  
      #parse params hash for values
      range = params[:range] || nil
      max_results = params[:max_results] || nil
      sort_order = params[:sortorder] || nil
      single_events = params[:singleevents] || nil
      timezone = params[:ctz] || nil
      
      #set up query string
      query_string += "q=#{CGI.escape(query)}" if query
      if range
        if not range.is_a? Hash or (range.size > 0 and (not range[:start].is_a? Time or not range[:end].is_a? Time))
          raise "The date range must be a hash including the :start and :end date values as Times"
        else
          date_range = ''
          if range.size > 0
            query_string += "&start-min=#{range[:start].xmlschema}&start-max=#{range[:end].xmlschema}"
          end
        end
      end
      query_string += "&max-results=#{max_results}" if max_results
      query_string += "&sortorder=#{sort_order}" if sort_order
      query_string += "&ctz=#{timezone.gsub(" ", "_")}" if timezone
      query_string += "&singleevents#{single_events}" if single_events
      if query_string
        events = calendar.service.send_get("http://www.google.com/calendar/feeds/#{calendar.id}/private/full?"+query_string)
        ret = []
        REXML::Document.new(events.read_body).root.elements.each("entry"){}.map do |entry|
          entry.attributes["xmlns:gCal"] = "http://schemas.google.com/gCal/2005"
          entry.attributes["xmlns:gd"] = "http://schemas.google.com/g/2005"
          entry.attributes["xmlns:app"] = "http://www.w3.org/2007/app"
          entry.attributes["xmlns"] = "http://www.w3.org/2005/Atom"
          event = Event.new(calendar)
          event.load("<?xml version='1.0' encoding='UTF-8'?>#{entry.to_s}")
          ret << event
        end
      end
      if params[:scope] == :first
        return ret[0]
      else
        return ret
      end
    end
    
    #Returns true if the event exists on the Google Calendar Service.
    def exists?
      return @exists
    end
  
    private 
    @exists = false
    @calendar = nil
    @xml = nil
    @etag = nil
    @recurrence = nil
    @deleted = false
    @edit_feed = ''
    
    def set_reminder(ele)
      ele.delete_element("gd:reminder")
      if @reminder
        e = ele.add_element("gd:reminder")
        used = false
        if @reminder[:minutes] 
          e.attributes['minutes'] = @reminder[:minutes] 
          used = true
        elsif @reminder[:hours] and not used
          e.attributes['hours'] = @reminder[:hours]
          used = true
        elsif @reminder[:days] and not used
          e.attributes['days'] = @reminder[:days]
        end
        if @reminder[:method] 
          e.attributes['method'] = @reminder[:method]
        else
          e.attributes['method'] = 'email'
        end
      else
        ele.delete_element("gd:reminder")
      end
    end
  end
end

